protocol/communities/manager.go

```diff
@@ -1370,65 +1371,144 @@ func (m *Manager) HandleCommunityEventsMessage(signer *ecdsa.PublicKey, message
+			pk, err := common.HexToPubkey(existingRequestToJoin.PublicKey)
+			if err != nil {
+				return updated, err
 			}
-		} else {
-			err := m.persistence.SaveRequestToJoin(requestToJoin)
+			err = m.persistence.SetRequestToJoinState(common.PubkeyToHex(pk), communityID, requestToJoin.State)
 			if err != nil {
-				return err
+				return updated, err
 			}
+			updated = true
+		}
+	} else {
+		err := m.persistence.SaveRequestToJoin(requestToJoin)
+		if err != nil {
+			return updated, err
 		}
-		return nil
 	}
+	return updated, nil
+}
 
-	for i := range community.config.EventsData.Events {
-		communityEvent := &community.config.EventsData.Events[i]
-		switch communityEvent.Type {
-		case protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_ACCEPT:
-			for signer, request := range communityEvent.AcceptedRequestsToJoin {
-				err := saveOrUpdateRequestToJoin(signer, request, RequestToJoinStateAccepted)
-				if err != nil {
-					return err
-				}
+func (m *Manager) handleCommunityEventRequestAccepted(community *Community, communityEvent *CommunityEvent) error {
+	requestToJoinState := RequestToJoinStateAccepted
+	if community.HasPermissionToSendCommunityEvents() {
+		// if we're an admin and we receive this admin event, we know the state is `pending`
+		requestToJoinState = RequestToJoinStateAcceptedPending
+	}
+
+	acceptedRequestsToJoin := make([]types.HexBytes, 0)
+
+	for signer, request := range communityEvent.AcceptedRequestsToJoin {
+		requestToJoin := &RequestToJoin{
+			PublicKey:   signer,
+			Clock:       request.Clock,
+			ENSName:     request.EnsName,
+			CommunityID: request.CommunityId,
+			State:       requestToJoinState,
+		}
+		requestToJoin.CalculateID()
+
+		if community.HasPermissionToSendCommunityEvents() {
+			existingRequestToJoin, err := m.persistence.GetRequestToJoin(requestToJoin.ID)
+			if err != nil && err != sql.ErrNoRows {
+				return err
+			}
+			if existingRequestToJoin.MarkedAsPendingByPrivilegedAccount() {
+				// the request is already in some pending state so we won't override it again
+				continue
 			}
+		}
 
-		case protobuf.CommunityEvent_COMMUNITY_REQUEST_TO_JOIN_REJECT:
-			for signer, request := range communityEvent.RejectedRequestsToJoin {
-				err := saveOrUpdateRequestToJoin(signer, request, RequestToJoinStateDeclined)
-				if err != nil {
-					return err
-				}
+		requestUpdated, err := m.saveOrUpdateRequestToJoin(signer, community.ID(), requestToJoin)
+		if err != nil {
+			return err
+		}
+
+		if requestUpdated {
+			// We only want to do this when `requestUpdated == true`, if the control node
+			// didn't know this request before, it means it just learned about it via this
+			// community event, which in turn means, we don't have any revealed addresses
+			// for this request yet, so we don't actually want to perform an
+			// accept action by the control node
+			acceptedRequestsToJoin = append(acceptedRequestsToJoin, requestToJoin.ID)
+		}
```

Comment:
The first one wins. But this can indeed cause a state where 2 admins look at different pending states until the ultimate state was confirmed by the control node.

To avoid that, we'd have to also store which account made which decision

Reply:
I was wondering about the scenario where two or more admin events come to control node before request to join:

1) acceptedRequestToJoin by adminA at timestamp 1
2) acceptedRequestToJoin by adminB at timestamp 2

at 2) `acceptedRequestsToJoin` will be populated with request that don't have revealed addresses. As a result, control node will reject the request before receiving addresses.

Reply:
So what I'd expect to happen in this case (and I'll add a test for this to verify this):

1. adminA decision reaches control node. Control node saves this request to join as it has seen it for the first time (at this point the request is in state acceptedPending
2. adminB's decision reaches control node. Control node sees that it already have a request with the same ID **and** it's in some pending state (acceptedPending or rejectedPending), so it'll ignore it altogether
3. Ultimately, control node will receive requestToJoin from user and perform the necessary checks. If there is already a request to join from an admin (meaning, the request exists in the database and is in any pending state coming from an admin), then it knows it can automatically accept (or reject) the request based on that

The important part here is that in 3) the request comes in via normal `HandleRequestToJoin()` while 1) and 2) are handled via `HandleCommunityEventsMessage`.

Reply:
> 2. adminB's decision reaches control node. Control node sees that it already have a request with the same ID and it's in some pending state (acceptedPending or rejectedPending), so it'll ignore it altogether

I can't find this part in the code, could you please point me where this check happens?

Reply:
Ah, I'm relying on this https://github.com/status-im/status-go/pull/3843/files#diff-0c22647108517d38489ae5faa65a914ade3be02ea7a7e37355a82739da9aeb69R1411 which doesn't check for pending state. It'll basically ignore any additional incoming decision from an admin, regardless of the state of existing request. Should probably add a check for pending states there

Reply:
> It'll basically ignore any additional incoming decision from an admin, regardless of the state of existing request

On the contrary, it will update the existing request if the new one (with higher clock value) comes :thinking: 

https://github.com/status-im/status-go/blob/bf60cd6a265495c677de9f0fce06e238670672d7/protocol/communities/manager.go#L1411-L1421

Reply:
> On the contrary, it will update the existing request if the new one (with higher clock value) comes 🤔

That's correct, but we expect it to be the same requestToJoin coming from different admins, so the clock will be the same no?

If you cancel your reques to join and send a new one, then it's in fact a new one which is fine (unless I'm missing something

Reply:
> so the clock will be the same no?

Oh, I missed that. I assumed the clock is altered once admin accepts/rejects the request. All fine then, thanks for patience.

> The first one wins. But this can indeed cause a state where 2 admins look at different pending states until the ultimate state was confirmed by the control node.

That's acceptable to me.

Reply:
Thank you for exploring this in detail! I'm going to add a test case for this regardless :)